actions/websocket_actions.test.jsx

```diff
@@ -930,21 +969,199 @@ describe('handlePluginEnabled/handlePluginDisabled', () => {
+            version: '0.2.0',
+            min_server_version: '5.12.0',
+            server: {
+                executables: {
+                    'linux-amd64': 'server/dist/plugin-linux-amd64',
+                    'darwin-amd64': 'server/dist/plugin-darwin-amd64',
+                    'windows-amd64': 'server/dist/plugin-windows-amd64.exe',
+                },
+            },
+            webapp: {
+                bundle_path: 'webapp/dist/main.js',
+            },
+        };
+
+        let getWebappPlugins;
+        let loadPluginsIfNecessary;
+
+        beforeEach(async () => {
+            console.log = jest.fn();
+            console.error = jest.fn();
+
+            const {Client4} = require('mattermost-redux/client'); // eslint-disable-line global-require
+            getWebappPlugins = Client4.getWebappPlugins;
+            loadPluginsIfNecessary = jest.requireActual('plugins').loadPluginsIfNecessary;
+
+            document.createElement = jest.fn();
+            document.getElementsByTagName = jest.fn();
+            document.getElementsByTagName.mockReturnValue([{
+                appendChild: jest.fn(),
+            }]);
+        });
+
+        test('should remove unused old plugins when socket reconnects', async () => {
+            const manifest = {
+                ...baseManifest,
+                id: 'com.mattermost.demo-2-plugin',
+            };
+
+            const initialize = jest.fn();
+            window.plugins = {
+                [manifest.id]: {
+                    initialize,
+                },
+            };
+            const mockScript = {};
+            document.createElement.mockReturnValue(mockScript);
+
+            expect(mockScript.onload).toBeUndefined();
+
+            // Enable plugin
+            handlePluginEnabled({data: {manifest}});
+            mockState.plugins.manifests = {[manifest.id]: manifest};
+
+            const dispatchReceivedArg = store.dispatch.mock.calls[0][0];
+            expect(dispatchReceivedArg.type).toBe(ActionTypes.ADD_WEBAPP_PLUGIN);
+            expect(dispatchReceivedArg.data).toBe(manifest);
+
+            expect(document.createElement).toHaveBeenCalledWith('script');
+            expect(document.createElement).toHaveBeenCalledTimes(1);
+
+            // Disable plugin with reload
+            mockState.entities.general.config.PluginsEnabled = 'true';
+            getWebappPlugins.mockImplementation(() => []);
+            await loadPluginsIfNecessary();
+            mockState.plugins.manifests = {};
+
+            const dispatchRemovedArg = store.dispatch.mock.calls[1][0];
+            expect(dispatchRemovedArg.type).toBe(ActionTypes.REMOVE_WEBAPP_PLUGIN);
             expect(dispatchRemovedArg.data).toBe(manifest);
+
+            // Assert disabling is idempotent
+            await loadPluginsIfNecessary();
+
+            expect(store.dispatch).toHaveBeenCalledTimes(3);
```

Comment:
I guess we are asserting here how many times an action was dispatched, but if we could confirm the last action then that would confirm that `loadPluginsIfNecessary` did not invoke an extra `ADD_WEBAPP_PLUGIN`

Reply:
I added a check that additional dispatch is not Add or Remove plugin